src/metrics/index.js

```diff
@@ -0,0 +1,244 @@
+      peerStats.stop()
+      this._peerStats.delete(idString)
+      this._oldPeers.set(idString, peerStats)
+    }
+  }
+
+  onMessage ({ remotePeer, protocol, direction, dataLength }) {
+    if (!this._running) return
+
+    const key = directionToEvent[direction]
+
+    let peerStats = this.forPeer(remotePeer)
+    if (!peerStats) {
+      peerStats = new Stats(initialCounters, this._options)
+      this._peerStats.set(remotePeer.toString(), peerStats)
+    }
+
+    // Protocol specific stats
+    if (protocol) {
+      let protocolStats = this.forProtocol(protocol)
+      if (!protocolStats) {
+        protocolStats = new Stats(initialCounters, this._options)
+        this._protocolStats.set(protocol, protocolStats)
+      }
+      protocolStats.push(key, dataLength)
+    // General stats
+    } else {
+      peerStats.push(key, dataLength)
+      this._globalStats.push(key, dataLength)
+    }
+  }
+
+  /**
+   * Replaces the `PeerId` string with the given `peerId`.
+   * If stats are already being tracked for the given `peerId`, the
+   * placeholder stats will be merged with the existing stats.
+   *
+   *
+   * @param {string} placeholder A peerId string
+   * @param {PeerId} peerId
+   */
+  updatePlaceholder (placeholder, peerId) {
+    if (!this._running) return
+    const placeholderStats = this.forPeer(placeholder)
+    const peerIdString = peerId.toString()
+    const existingStats = this.forPeer(peerId)
+    let mergedStats = placeholderStats
+
+    // If we already have stats, merge the two
+    if (existingStats) {
+      // If existing, merge
+      mergedStats = Metrics.mergeStats(existingStats, mergedStats)
+      // Attempt to delete from the old peers list just in case it was tracked there
+      this._oldPeers.delete(peerIdString)
+    }
+
+    this._peerStats.delete(placeholder.toString())
+    this._peerStats.set(peerIdString, mergedStats)
+    mergedStats.start()
+  }
+
+  /**
+   * Tracks data running through a given Duplex Iterable `stream`. If
+   * the `peerId` is not provided, a placeholder string will be created and
+   * returned. This allow lazy tracking of a peer when the peer is not yet known.
+   * When the `PeerId` is known, `Metrics.updatePlaceholder` should be called
+   * with the placeholder string returned from here, and the known `PeerId`.
+   *
+   * @param {Object} options
+   * @param {{ sink: function(*), source: function() }} options.stream A duplex iterable stream
+   * @param {PeerId} [options.peerId] The id of the remote peer that's connected
+   * @param {string} [options.protocol] The protocol the stream is running
+   * @returns {string} The peerId string or placeholder string
+   */
+  trackStream ({ stream, remotePeer, protocol }) {
+    const metrics = this
+    const _source = stream.source
+    stream.source = (async function * () {
+      for await (const chunk of _source) {
+        metrics.onMessage({ remotePeer, protocol, direction: 'in', dataLength: chunk.length })
```

Comment:
it would be nice to use [streaming-iterables#tap](https://github.com/bustle/streaming-iterables#tap) with pipe here